COMPACT_DENYLIST_FORMAT.md

```diff
@@ -0,0 +1,457 @@
+
+```
+echo /ipfs/QmecDgNqCRirkc3Cjz9eoRBNwXGckJ9WvTdmY16HP88768 >> ~/.config/ipfs/custom.deny
+```
+
+We conciously avoid defining any other API other than expecting
+implementations to honor blocking what is on the denylist and act accordingly
+when it is updated. CLI commands or API endpoint to modify list items etc. are
+outside the scope of this spec. Implementations how much information to
+provide to users when a request for an IPFS object is blocked.
+
+As a last note, if we take Kubo and the go-ipfs stack as the reference IPFS
+implementation, we expect the blocking-layer (that is, the introduction of the
+logic that decides whether an item is blocked or not), to happen cleanly at
+the `NameSystem`, `path.Resolver` and `BlockService` interfaces (IPNS, IPFS
+Path and CID blocks respectively).
+
+This specification corresponds to V1 of the compact list format. We have
+limited the number of features and extensions to a minimum to start working
+with, leaving some ideas on the table and the door open to develop the format
+in future versions.
+
+## Specification
+
+### Denylist file extension, locations and order
+
+While not pertaining to the denylist format itself, we introduce the following conventions about denylist files when they are stored in the local filesystem:
+
+- Denylist files are named with the extension `.deny`.
+- Implementations should look in `/etc/ipfs/denylists/` and
+  `$XDG_CONFIG_HOME/ipfs/denylists/` for denylist files.
+- Denylist files are processed in alphabetical order so that rules from later
+  denylists override rules from earlier denylists on conflict.
+
+### Denylist format
+
+#### Summary
+
+The following example showcases the features and syntax of a compact denylist:
+
+```
+version: 1
+name: IPFSCorp blocking list
+description: A collection of bad things we have found in the universe
+author: abuse-ipfscorp@example.com
+hints:
+  hint: value
+  hint2: value2
+---
+# Blocking by CID - blocks wrapped multihash.
+# Does not block subpaths.
+/ipfs/bafybeihvvulpp4evxj7x7armbqcyg6uezzuig6jp3lktpbovlqfkuqeuoq
+
+# Block all subpaths
+/ipfs/QmdWFA9FL52hx3j9EJZPQP1ZUH8Ygi5tLCX2cRDs6knSf8/*
+
+# Block some subpaths (equivalent rules)
+/ipfs/Qmah2YDTfrox4watLCr3YgKyBwvjq8FJZEFdWY6WtJ3Xt2/test*
+/ipfs/QmTuvSQbEDR3sarFAN9kAeXBpiBCyYYNxdxciazBba11eC/test/*
+
+# Block some subpaths with exceptions: last-matching-rule wins (!)
+/ipfs/QmUboz9UsQBDeS6Tug1U8jgoFkgYxyYood9NDyVURAY9pK/blocked*
+!/ipfs/QmUboz9UsQBDeS6Tug1U8jgoFkgYxyYood9NDyVURAY9pK/blockednot
+!/ipfs/QmUboz9UsQBDeS6Tug1U8jgoFkgYxyYood9NDyVURAY9pK/blocked/not
+!/ipfs/QmUboz9UsQBDeS6Tug1U8jgoFkgYxyYood9NDyVURAY9pK/blocked/exceptions*
+
+# Block IPNS domain name
+/ipns/domain.example
+
+# Block IPNS domain name and path
+/ipns/domain2.example/path
+
+# Block IPNS key - blocks wrapped multihash.
+/ipns/k51qzi5uqu5dhmzyv3zac033i7rl9hkgczxyl81lwoukda2htteop7d3x0y1mf
+
+# Block all mime types with exceptions
+/mime/image/*
+!/mime/image/gif
```

Comment:
Is this something that a denylist format needs to support? It feels like it belongs in a local configuration format rather than with something that will get distributed. What's more, media type information is unreliable and rarely available on IPFS as it tends to rely on sniffing.

Reply:
right, I'd break this into two questions that should ideally be decided before pouring cement:
1.) is the system layered enough that local configs and/or processing of lists after merging is a better place for this to economize at higher distro/merging layer, and
2.) is this a required capability or an optional/extension feature (to allow economizing compute at any layer if implementor needs to avoid sniffing anything)

Reply:
Mime-blocking will likely have to rely on sniffing and likely on stuff that can be interpreted as a file (so unixfs). The concern is warranted. It would also mostly apply to gateways, as it is the main thing turning blocks into files again.

I would be fine striking this off. If someone wants a text-only gateway then it is probably better done with nginx.

Reply:
Removed mime blocking from spec.

Reply:
@hsanjuan should we remove it from excamples here too?

Reply:
Assuming yes. Removed remaining mentions in 7233af2
We can always add them in future IPIP.